launcher-implementation/src/main/scala/xsbt/boot/Launch.scala

```diff
@@ -314,8 +314,8 @@ class Launch private[xsbt] (val bootDirectory: File, val lockBoot: Boolean, val
       def mainClass: Class[T] forSome { type T <: xsbti.AppMain } = entryPoint.asSubclass(AppMainClass)
       def newMain(): xsbti.AppMain = {
         if (ServerApplication.isServerApplication(entryPoint)) ServerApplication(this)
-        else if (PlainApplication.isPlainApplication(entryPoint)) PlainApplication(entryPoint)
         else if (AppMainClass.isAssignableFrom(entryPoint)) mainClass.newInstance
+        else if (PlainApplication.isPlainApplication(entryPoint)) PlainApplication(entryPoint)
```

Comment:
I remember this throwing classcastexceptions in that past. If it weren't for the tests, I wouldn't believe it.